realm/src/main/java/io/realm/internal/async/RealmThreadPoolExecutor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.async;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Custom thread pool settings, instances of this executor can be paused, and resumed.
+ */
+public class RealmThreadPoolExecutor extends ThreadPoolExecutor {
+    // reduce context switch by using a number of thread proportionate to the number of cores
+    private static final int CORE_POOL_SIZE = Runtime.getRuntime().availableProcessors() * 2 + 1;
```

Comment:
This number always seemed a bit magical to me. I know that AsyncTask does the same, but we still have more threads than available processors so context switches are bound to happen anyway.